profiles/audit_log.py

```diff
@@ -0,0 +1,100 @@
+import json
+import logging
+import threading
+from datetime import datetime, timezone
+
+from django.conf import settings
+
+from .utils import get_current_service, get_current_user, get_original_client_ip
+
+_thread_locals = threading.local()
+
+
+def _should_audit(model):
+    if hasattr(model, "audit_log") and model.audit_log:
+        return True
+    return False
+
+
+def add_loggable(action, instance):
+    if (
+        settings.AUDIT_LOGGING_ENABLED
+        and _should_audit(instance.__class__)
+        and instance.pk
+    ):
+        current_time = datetime.now(tz=timezone.utc)
+
+        if not hasattr(_thread_locals, "loggables"):
+            _thread_locals.loggables = list()
+        _thread_locals.loggables.append((current_time, action, instance))
+
+
+def _resolve_role(current_user, profile):
+    if profile.user == current_user:
+        return "OWNER"
+    elif current_user is not None:
+        return "ADMIN"
+    else:
+        return "SYSTEM"
+
+
+def _format_user_data(audit_event, field_name, user):
+    if user:
+        audit_event[field_name]["user_id"] = (
+            str(user.uuid) if hasattr(user, "uuid") else None
+        )
+        if settings.AUDIT_LOG_USERNAME:
+            audit_event[field_name]["user_name"] = (
+                user.username if hasattr(user, "username") else None
```

Comment:
This is a question also for @vikoivun :
We have an ASVS requirement 7.3.1: encode user supplied data to prevent log injection. @vikoivun can chime in whether this is a realistic problem. (The data flow seems to be: python dict -> json.dumps -> sent to log system).
https://github.com/OWASP/ASVS/blob/master/4.0/en/0x15-V7-Error-Logging.md#v73-log-protection-requirements

If this is not safe here, the encoding should preferably made in a systematic way, not field-by-field.

Reply:
I now realize the username field is system-generated, so this probably is irrevelant.

Reply:
Still it is a good idea to check how Elasticsearch handles arbitrary data. I'd guess it is pretty tolerant of almost anything. How Kibana then handles stuff is another matter.

Reply:
Django puts this restriction to the username: "150 characters or fewer. Letters, digits and @/./+/-/_ only.".